tagstore/backends/django.py

```diff
@@ -0,0 +1,155 @@
+        for entity in entity_ids:
+            (em, _created) = EntityModel.objects.get_or_create(
+                entity_type=entity.entity_type,
+                external_id=entity.external_id,
+            )
+            em.tags.add(tag_model)
+
+    def remove_tag_from(self, tag: Tag, *entity_ids: EntityId) -> None:
+        """
+        Remove the specified tag from the specified entity/entities
+
+        Will be a no-op if the entities do not have that tag.
+        """
+        tag = TagModel.objects.get(taxonomy_id=tag.taxonomy_uid, tag=tag.tag)
+        # This could be optimized to a single DB query, but that's probably not necessary
+        for eid in entity_ids:
+            try:
+                EntityModel.objects.get(entity_type=eid.entity_type, external_id=eid.external_id).tags.remove(tag)
+            except EntityModel.DoesNotExist:
+                pass
+
+    def get_tags_applied_to(self, *entity_ids: EntityId) -> TagSet:
+        """ Get the set of unique tags applied to any of the specified entity IDs """
+        entity_filter = None
+        for eid in entity_ids:
+            q = Q(entity_type=eid.entity_type) & Q(external_id=eid.external_id)
+            if entity_filter is None:
+                entity_filter = q
+            else:
+                entity_filter = entity_filter | q
+        entities = EntityModel.objects.filter(entity_filter)
+        tags = TagModel.objects.filter(entity__id__in=Subquery(entities.values('id')))
+        tags_found = set()
+        for tag in tags:
+            tags_found.add(Tag(taxonomy_uid=tag.taxonomy_id, tag=tag.tag))
+        return tags_found
+
+    # Searching Entities ##########################
+
+    def get_entities_tagged_with_all(
+        self,
+        tags: TagSet,
+        entity_types: Optional[List[str]] = None,
+        external_id_prefix: Optional[str] = None,
+        entity_ids: Optional[List[EntityId]] = None,  # use this to filter a list of entity IDs by tag
+        include_child_tags=True,  # For hierarchical taxonomies, include child tags
+                                  # (e.g. search for "Animal" will return results tagged only with "Dog")
+    ) -> Iterator[EntityId]:
+
+        if not tags:
+            raise ValueError("tags must contain at least one Tag")
+
+        entities = EntityModel.objects
+
+        if include_child_tags:
+            # Convert the set of tags to a set of materialized paths:
+            tags_filter = Q()
+            for tag in tags:
+                tags_filter = tags_filter | (Q(taxonomy_id=tag.taxonomy_uid) & Q(tag=tag.tag))
+            paths = TagModel.objects.filter(tags_filter).values_list('path', flat=True)
+            for path in paths:
+                entities = entities.filter(tags__path__startswith=path)
+        else:
+            for tag in tags:
+                entities = entities.filter(tags__taxonomy_id=tag.taxonomy_uid, tags__tag=tag.tag)
```

Comment:
If the tag provided in the `TagSet` doesn't exist in the database, then I get a `TypeError` below, because no filters are applied here.  E.g.
```python
dogtag = Tag(1, 'dog')
list(self.tagstore.get_entities_tagged_with_all({dogtag}))
# (0.000) SELECT "tagstore_tag"."path" FROM "tagstore_tag" WHERE ("tagstore_tag"."taxonomy_id" = 1 AND "tagstore_tag"."tag" = 'dog') ORDER BY "tagstore_tag"."tag" ASC; args=(1, 'dog')
# Traceback (most recent call last):
#   File "<console>", line 1, in <module>
#   File "blockstore/tagstore/backends/django.py", line 154, in get_entities_tagged_with_all
#     for e in entities:
# TypeError: 'Manager' object is not iterable
```

Reply:
Thanks, fixed in 67c54b2efba10a0d53f9fba957eac3e1e64f1fcd by changing the first line to `entities = EntityModel.objects.all()`

I didn't both writing a test case for this, since manually constructing `Tag` objects that don't exist in the DB is not something the API is supposed to support.